src/peergos/shared/user/UserContext.java

```diff
@@ -1650,13 +1650,22 @@ private static void ensureAllowedToShare(FileWrapper file, String ourname, boole
     }
 
     @JsMethod
-    public CompletableFuture<List<FileWrapper>> getFiles(List<SharedItem> pointers) {
-        return Futures.combineAllInOrder(pointers.stream()
-                .map(s -> network.getFile(s.cap, s.owner))
-                .collect(Collectors.toList()))
-                .thenApply(res -> res.stream()
-                        .flatMap(Optional::stream)
-                        .collect(Collectors.toList()));
+    public CompletableFuture<List<Pair<SharedItem,FileWrapper>>> getFiles(List<SharedItem> pointers) {
+        List<CompletableFuture<Optional<?>>> futures = pointers.stream().map(s -> network.getFile(s)).collect(Collectors.toList());
+        CompletableFuture<List<Pair<SharedItem, FileWrapper>>> identity = CompletableFuture.completedFuture(Collections.emptyList());
+        return futures.stream().reduce(identity,
+                (a, b) -> b.thenCompose(opt ->
+                        a.thenApply(set -> {
+                            ArrayList<Pair<SharedItem, FileWrapper>> combined = new ArrayList<>(set.size() + 1);
```

Comment:
this would make the algorithm O(n^2) in the size of pointers

Reply:
That means combineAllInOrder suffers from this also.

Reply:
No it doesn't. The n squared comes from creating a new list at each step. First it has 1 element, then 2 then 3 ... 1+2+3...+n = n(n+1)/2